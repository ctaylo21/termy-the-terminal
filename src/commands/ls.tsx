import React from 'react';
import get from 'lodash/get';
import has from 'lodash/has';
import { getInternalPath } from './utilities/index';
import LsResult from '../components/LsResult';
import { AutoCompleteResponse, CommandResponse, FileSystem } from '../index';

export interface LsResultType {
  [index: string]: {
    type: 'FOLDER' | 'FILE';
  };
}

function getTarget(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string,
): FileSystem {
  const internalPath = getInternalPath(currentPath, targetPath);

  if (internalPath === '/' || !internalPath) {
    return fileSystem;
  } else if (has(fileSystem, internalPath)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const target = get(fileSystem, internalPath);
    if (target.type === 'FILE') {
      const [fileName] = internalPath.split('.').slice(-1);
      return { [fileName]: target };
    } else {
      if (target.children) {
        return target.children;
      } else {
        throw new Error('Nothing to show here');
      }
    }
  }

  throw new Error('Target folder does not exist');
}

/**
 * Takes an internally formatted filesystem and formats it into the
 * expected format for an ls command. Optionally takes a function to apply
 * to the intiial result to filter out certain items.
 *
 * @param directory {object} - internally formatted filesystem
 * @param filterFn {function} - optional fn to filter certain items
 */
function buildLsFormatDirectory(
  fileSystem: FileSystem,
  filterFn: (item: LsResultType) => boolean = (): boolean => true,
): LsResultType {
  return Object.assign(
    {},
    ...Object.keys(fileSystem)
      .map((item) => ({
        [fileSystem[item].type === 'FILE'
          ? `${item}.${fileSystem[item].extension}`
          : item]: {
          type: fileSystem[item].type,
        },
      }))
      .filter(filterFn),
  );
}

/**
 * Given a fileysystem, lists all items for a given directory
 *
 * @param fileSystem {object} - filesystem to ls upon
 * @param currentPath {string} - current path within filesystem
 * @param targetPath {string} - path to list contents within
 * @returns Promise<object> - resolves with contents of given path
 */
export default function ls(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath = '',
): Promise<CommandResponse> {
  return new Promise((resolve, reject): void => {
    let targetFolderContents;
    try {
      targetFolderContents = getTarget(fileSystem, currentPath, targetPath);
    } catch (e) {
      return reject(e.message);
    }

    resolve({
      commandResult: (
        <LsResult lsResult={buildLsFormatDirectory(targetFolderContents)} />
      ),
    });
  });
}

/**
 * Given a fileysystem, current path, and target, list the items in the desired
 * folder that start with target string
 *
 * @param fileSystem {object} - filesystem to ls upon
 * @param currentPath {string} - current path within filesystem
 * @param target {string} - string to match against (maybe be path)
 * @returns Promise<object> - resolves with contents that match target in path
 */
function lsAutoComplete(
  fileSystem: FileSystem,
  currentPath: string,
  target: string,
): Promise<AutoCompleteResponse> {
  return new Promise((resolve): void => {
    // Default to searching in currenty directory with simple target
    // that contains no path
    let autoCompleteMatch = target;
    let targetPath = '';

    // Handle case where target is a nested path and
    // we need to pull off last part of path to match against
    const pathParts = target.split('/');
    if (pathParts.length > 1) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      autoCompleteMatch = pathParts.pop()!;
      targetPath = pathParts.join('/');
    }

    let targetFolderContents;
    try {
      targetFolderContents = getTarget(fileSystem, currentPath, targetPath);
    } catch (e) {
      return resolve({ commandResult: undefined });
    }

    const matchFilterFn = (item: LsResultType): boolean =>
      Object.keys(item)[0].startsWith(autoCompleteMatch);

    resolve({
      commandResult: buildLsFormatDirectory(
        targetFolderContents,
        matchFilterFn,
      ),
    });
  });
}

export { ls, lsAutoComplete };
