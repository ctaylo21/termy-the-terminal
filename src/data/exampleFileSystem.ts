const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          video1: {
            type: 'FILE',
          },
        },
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
};

export default exampleFileSystem;
