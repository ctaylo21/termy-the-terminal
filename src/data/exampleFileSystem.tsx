import React from 'react';
import { FileSystem } from '../index';
import dogImg from '../../src/images/dog.png';

interface BlogPostProps {
  date: string;
  content: string;
}

export const BlogPost: React.FC<BlogPostProps> = ({
  date,
  content,
}): JSX.Element => (
  <>
    <h3>{date}</h3>
    <p>{content}</p>
  </>
);

const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          file2: {
            type: 'FILE',
            content: 'Contents of file 2',
            extension: 'txt',
          },
        },
      },
      dog: {
        type: 'FILE',
        content: dogImg,
        extension: 'png',
      },
      file1: {
        type: 'FILE',
        content: 'Contents of file 1',
        extension: 'txt',
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
  file3: {
    type: 'FILE',
    content: 'Contents of file 3',
    extension: 'txt',
  },
  blog: {
    type: 'FILE',
    content: <BlogPost date="3/22" content="Today is a good day" />,
    extension: 'txt',
  },
};

export default exampleFileSystem;
