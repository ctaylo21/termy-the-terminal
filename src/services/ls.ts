import get from 'lodash/get';
import { convertPathToInternalFormat } from './utilities/index';

function getTargetFolder(
  fileSystem: FileSystem,
  targetPath: string,
): FileSystem | null {
  const internalPath = convertPathToInternalFormat(targetPath);
  return targetPath === '/'
    ? fileSystem
    : (get(fileSystem, internalPath) as TerminalFolder).children;
}

/**
 * Given a fileysystem, lists all items for a given directory
 *
 * @param fileSystem {object} - filesystem to ls upon
 * @param targetPath {string} - path to list contents within
 * @returns Promise<string> - resolves with contents of given path
 */
export default function ls(
  fileSystem: FileSystem,
  targetPath: string,
): Promise<LsResultType> {
  return new Promise(
    (resolve): void => {
      const externalFormatDir: LsResultType = {};
      const targetFolderContents = getTargetFolder(fileSystem, targetPath);

      for (let key in targetFolderContents) {
        externalFormatDir[key] = {
          type: targetFolderContents[key].type,
        };
      }
      resolve(externalFormatDir as LsResultType);
    },
  );
}
