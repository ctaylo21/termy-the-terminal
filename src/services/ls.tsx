import React from 'react';
import get from 'lodash/get';
import has from 'lodash/has';
import { getInternalPath } from './utilities/index';
import LsResult from '../components/LsResult';

function getTargetFolder(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string,
): FileSystem | null {
  const internalPath = getInternalPath(currentPath, targetPath);

  if (internalPath === '/' || !internalPath) {
    return fileSystem;
  } else if (has(fileSystem, internalPath)) {
    return (get(fileSystem, internalPath) as TerminalFolder).children;
  }

  throw new Error('Target folder does not exist');
}

/**
 * Given a fileysystem, lists all items for a given directory
 *
 * @param fileSystem {object} - filesystem to ls upon
 * @param currentPath {string} - current path within filesystem
 * @param targetPath {string} - path to list contents within
 * @returns Promise<object> - resolves with contents of given path
 */
export default function ls(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string = '',
): Promise<ServiceResponse> {
  return new Promise(
    (resolve, reject): void => {
      const externalFormatDir: LsResultType = {};

      let targetFolderContents;
      try {
        targetFolderContents = getTargetFolder(
          fileSystem,
          currentPath,
          targetPath,
        );
      } catch (e) {
        reject(e.message);
      }

      for (let key in targetFolderContents) {
        const lsKey =
          targetFolderContents[key].type === 'FILE'
            ? `${key}.${targetFolderContents[key].extension}`
            : key;

        externalFormatDir[lsKey] = {
          type: targetFolderContents[key].type,
        };
      }
      resolve({
        serviceResult: <LsResult lsResult={externalFormatDir} />,
      });
    },
  );
}
