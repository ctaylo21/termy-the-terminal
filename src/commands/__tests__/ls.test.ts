import { ls, lsAutoComplete } from '../ls';
import exampleFileSystem from '../../data/exampleFileSystem';
import { render } from '@testing-library/react';

describe('ls suite', (): void => {
  test('root with no directory', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('docs');
    expect(container.innerHTML).toContain('home');
  });

  test('nested path with no directory', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/home');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('user');
    expect(container.innerHTML).toContain('videos');
    expect(container.innerHTML).toContain('file1.txt');
  });

  test('relative path from root', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/', '/home');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('user');
    expect(container.innerHTML).toContain('videos');
    expect(container.innerHTML).toContain('file1.txt');
  });

  test('relative path from nested path', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/home', 'user');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('test');
  });

  test('root with dotdot', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/', '..');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('home');
    expect(container.innerHTML).toContain('docs');
  });

  test('nestd path with dotdot', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/home', '../home/user/..');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('user');
    expect(container.innerHTML).toContain('videos');
    expect(container.innerHTML).toContain('file1.txt');
  });

  test('should reject if invalid target given', async (): Promise<void> => {
    return expect(ls(exampleFileSystem, '/invalid')).rejects.toMatchSnapshot();
  });

  test('empty path from nested location', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/home/user', '');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('test');
  });

  test('should return file if given', async (): Promise<void> => {
    const lsResult = await ls(exampleFileSystem, '/', 'file4.txt');
    const { container } = render(lsResult.commandResult as JSX.Element);

    expect(container.innerHTML).toContain('file4.txt');
  });

  test('should return message if directory and empty', async (): Promise<
    void
  > => {
    return expect(
      ls(exampleFileSystem, '/home', 'user/test'),
    ).rejects.toMatchSnapshot();
  });

  describe('auto complete', (): void => {
    test('empty value', async (): Promise<void> => {
      const { commandResult } = await lsAutoComplete(
        exampleFileSystem,
        '/home/user',
        '',
      );

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(Object.keys(commandResult!)).toContain('test');
    });

    test('should filter single level target', async (): Promise<void> => {
      const { commandResult } = await lsAutoComplete(
        exampleFileSystem,
        '/',
        'fi',
      );

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('file3.txt');
      expect(autoCompleteValues).toContain('file4.txt');
      expect(autoCompleteValues).not.toContain('blog');
      expect(autoCompleteValues).not.toContain('docs');
      expect(autoCompleteValues).not.toContain('home');
    });

    test('invalid path should return nothing', async (): Promise<void> => {
      const lsResult = await lsAutoComplete(exampleFileSystem, '/bad/path', '');

      expect(lsResult.commandResult).toBeUndefined();
    });

    test('relative path', async (): Promise<void> => {
      const { commandResult } = await lsAutoComplete(
        exampleFileSystem,
        '/',
        'home/fi',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('file1.txt');
      expect(autoCompleteValues).toContain('file5.txt');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('dog.png');
    });

    test('relative path with dotdot', async (): Promise<void> => {
      const { commandResult } = await lsAutoComplete(
        exampleFileSystem,
        '/',
        'home/../home/fi',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('file1.txt');
      expect(autoCompleteValues).toContain('file5.txt');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('dog.png');
    });

    test('absolute path', async (): Promise<void> => {
      const { commandResult } = await lsAutoComplete(
        exampleFileSystem,
        '/',
        '/home/d',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('dog.png');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('file1.txt');
      expect(autoCompleteValues).not.toContain('file5.txt');
    });

    test('absolute path with ..', async (): Promise<void> => {
      const { commandResult } = await lsAutoComplete(
        exampleFileSystem,
        '/',
        '/home/user/../d',
      );
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const autoCompleteValues = Object.keys(commandResult!);

      expect(autoCompleteValues).toContain('dog.png');
      expect(autoCompleteValues).not.toContain('user');
      expect(autoCompleteValues).not.toContain('videos');
      expect(autoCompleteValues).not.toContain('file1.txt');
      expect(autoCompleteValues).not.toContain('file5.txt');
    });
  });
});
