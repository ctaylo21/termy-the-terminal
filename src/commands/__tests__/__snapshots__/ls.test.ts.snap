// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ls suite should reject if invalid target given 1`] = `"Target folder does not exist"`;

exports[`ls suite should return message if directory and empty 1`] = `"Nothing to show here"`;
